File: code/07-type_systems/example_7.59-60/Best.java
// Examples 7.59 and 7.60 (Figure 7.6)

interface Chooser<T> {
    public Boolean better(T a, T b);
}

class Arbiter<T> {
    T bestSoFar;
    Chooser<? super T> comp;

    public Arbiter(Chooser<? super T> c) {
        comp = c;
    }
    public void consider(T t) {
        if (bestSoFar == null || comp.better(t, bestSoFar)) bestSoFar = t;
    }
    public T best() {
        return bestSoFar;
    }
}

class CaseSensitive implements Chooser<String> {
    public Boolean better(String a, String b) {
        return a.compareTo(b) < 1;
    }
}

class CaseInsensitive implements Chooser<Object> {
    public Boolean better(Object a, Object b) {
        return a.toString().compareToIgnoreCase(b.toString()) < 1;
    }
}

public class Best {
    public static void main(String args[]) {
        Arbiter<String> csNames = new Arbiter<String>(new CaseSensitive());
        csNames.consider(new String("Apple"));
        csNames.consider(new String("aardvark"));
        System.out.println(csNames.best());         // Apple

        Arbiter<String> ciNames = new Arbiter<String>(new CaseInsensitive());
        ciNames.consider(new String("Apple"));
        ciNames.consider(new String("aardvark"));
        System.out.println(ciNames.best());         // aardvark
    }
}

File: code/07-type_systems/example_7.66-67/Best.cs
// Examples 7.66 and 7.67 (Figure 7.8)

using System;

interface Chooser<in T> {
    bool better(T a, T b);
}

class Arbiter<T> {
    T bestSoFar;
    Chooser<T> comp;
    bool initialized;

    public Arbiter(Chooser<T> c) {
        comp = c;
        bestSoFar = default(T);
        initialized = false;
    }
    public void Consider(T t) {
        if (!initialized || comp.better(t, bestSoFar)) bestSoFar = t;
        initialized = true;
    }
    public T Best() {
        return bestSoFar;
    }
}

class CaseSensitive : Chooser<String> {
    public bool better(String a, String b) {
        return String.CompareOrdinal(a, b) < 1;
    }
}

class CaseInsensitive : Chooser<Object> {
    public bool better(Object a, Object b) {
        return String.Compare(a.ToString(), b.ToString(), false) < 1;
    }
}

public class Best {
    public static void Main(String[] args)
    {
        Arbiter<String> csNames = new Arbiter<String>(new CaseSensitive());
        csNames.Consider("Apple");
        csNames.Consider("aardvark");
        Console.WriteLine(csNames.Best());      // Apple

        Arbiter<String> ciNames = new Arbiter<String>(new CaseInsensitive());
        ciNames.Consider("Apple");
        ciNames.Consider("aardvark");
        Console.WriteLine(ciNames.Best());      // aardvark

        Arbiter<int> OKinCsharp;
    }
}
